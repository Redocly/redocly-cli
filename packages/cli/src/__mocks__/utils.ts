export const getFallbackEntryPointsOrExit = jest.fn((entrypoints) => entrypoints.map(() => ({ path: '' })));
export const getTotals = jest.fn(() => ({ errors: 0 }));
export const dumpBundle = jest.fn(() => '');
export const slash = jest.fn();
export const pluralize = jest.fn();
export const getExecutionTime = jest.fn();
export const printExecutionTime = jest.fn();
export const printUnusedWarnings = jest.fn();
export const printLintTotals = jest.fn();
export const getOutputFileName = jest.fn(() => ({ outputFile: 'test.yaml', ext: 'yaml' }));
export const handleError = jest.fn();
