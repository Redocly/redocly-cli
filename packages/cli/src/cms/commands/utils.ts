import { pause } from '@redocly/openapi-core';

/**
 * This function retries an operation until a condition is met or a timeout is exceeded.
 * If the condition is not met within the timeout, an error is thrown.
 * @operation The operation to retry.
 * @condition The condition to check after each operation result. Return false to continue retrying. Return true to stop retrying.
 *            If not provided, the first result will be returned.
 * @param onConditionNotMet Will be called with the last result right after checking condition and before timeout and retrying.
 * @param onRetry Will be called right before retrying operation with the last result before retrying.
 * @param startTime The start time of the operation. Default is the current time.
 * @param retryTimeoutMs The maximum time to retry the operation. Default is 10 minutes.
 * @param retryIntervalMs The interval between retries. Default is 5 seconds.
 */
export async function retryUntilConditionMet<T>({
  operation,
  condition,
  onConditionNotMet,
  onRetry,
  startTime = Date.now(),
  retryTimeoutMs = 600000, // 10 min
  retryIntervalMs = 5000, // 5 sec
}: {
  operation: () => Promise<T>;
  condition?: ((result: T) => boolean) | null;
  onConditionNotMet?: (lastResult: T) => void;
  onRetry?: (lastResult: T) => void | Promise<void>;
  startTime?: number;
  retryTimeoutMs?: number;
  retryIntervalMs?: number;
}): Promise<T> {
  async function attempt(): Promise<T> {
    const result = await operation();

    if (!condition) {
      return result;
    }

    if (condition(result)) {
      return result;
    } else if (Date.now() - startTime > retryTimeoutMs) {
      throw new Error('Timeout exceeded');
    } else {
      onConditionNotMet?.(result);
      await pause(retryIntervalMs);
      await onRetry?.(result);
      return attempt();
    }
  }

  return attempt();
}
