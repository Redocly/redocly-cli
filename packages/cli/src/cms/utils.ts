export class DeploymentError extends Error {}
