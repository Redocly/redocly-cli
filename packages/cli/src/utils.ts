import { basename, dirname, extname, join, resolve } from 'path';
import { blue, gray, green, red, yellow } from 'colorette';
import { performance } from "perf_hooks";
import * as glob from 'glob-promise';
import * as fs from 'fs';
import * as path from 'path';
import * as readline from 'readline';
import { Writable } from 'stream';
import {
  BundleOutputFormat,
  Config,
  LintConfig,
  ResolveError,
  YamlParseError,
  parseYaml,
  stringifyYaml
} from '@redocly/openapi-core';
import { Totals, outputExtensions, Entrypoint } from './types';

export async function getFallbackEntryPointsOrExit(argsEntrypoints: string[] | undefined, config: Config): Promise<Entrypoint[]> {
  const  { apis } = config;
  const shouldFallbackToAllDefinitions = !isNotEmptyArray(argsEntrypoints) && apis && Object.keys(apis).length > 0;
  const res = shouldFallbackToAllDefinitions
    ? Object.entries(apis).map(([alias, { root }]) => ({
        path: resolve(getConfigDirectory(config), root),
        alias,
      }))
    : (await expandGlobsInEntrypoints(argsEntrypoints!, config));
  if (!isNotEmptyArray(res)) {
    process.stderr.write('error: missing required argument `entrypoints`.\n');
    process.exit(1);
  }
  return res;
}

function getConfigDirectory(config: Config) {
  return config.configFile ? dirname(config.configFile) : process.cwd();
}

function isNotEmptyArray<T>(args?: T[]): boolean {
  return Array.isArray(args) && !!args.length;
}

function getAliasOrPath(config: Config, aliasOrPath: string): Entrypoint {
  return config.apis[aliasOrPath]
    ? { path: config.apis[aliasOrPath]?.root, alias: aliasOrPath }
    : { path: aliasOrPath };
}

async function expandGlobsInEntrypoints(args: string[], config: Config) {
  return (await Promise.all((args as string[]).map(async aliasOrPath => {
    return glob.hasMagic(aliasOrPath)
      ? (await glob(aliasOrPath)).map((g: string) => getAliasOrPath(config, g))
      : getAliasOrPath(config, aliasOrPath);
  }))).flat();
}

export function getExecutionTime(startedAt: number) {
  return process.env.NODE_ENV === 'test'
    ? '<test>ms'
    : `${Math.ceil(performance.now() - startedAt)}ms`;
}

export function printExecutionTime(commandName: string, startedAt: number, entrypoint: string) {
  const elapsed = getExecutionTime(startedAt);
  process.stderr.write(gray(`\n${entrypoint}: ${commandName} processed in ${elapsed}\n\n`));
}

export function pathToFilename(path: string, pathSeparator: string) {
  return path
    .replace(/~1/g, '/')
    .replace(/~0/g, '~')
    .replace(/^\//, '')
    .replace(/\//g, pathSeparator);
}

export class CircularJSONNotSupportedError extends Error {
  constructor(public originalError: Error) {
    super(originalError.message);
    // Set the prototype explicitly.
    Object.setPrototypeOf(this, CircularJSONNotSupportedError.prototype);
  }
}

export function dumpBundle(obj: any, format: BundleOutputFormat, dereference?: boolean): string {
  if (format === 'json') {
    try {
      return JSON.stringify(obj, null, 2);
    } catch (e) {
      if (e.message.indexOf('circular') > -1) {
        throw new CircularJSONNotSupportedError(e);
      }
      throw e;
    }
  } else {
    return stringifyYaml(obj, {
      noRefs: !dereference,
    });
  }
}

export function saveBundle(filename: string, output: string) {
  fs.mkdirSync(path.dirname(filename), { recursive: true });
  fs.writeFileSync(filename, output);
}

export async function promptUser(query: string, hideUserInput = false): Promise<string> {
  return new Promise((resolve) => {
    let output: Writable = process.stdout;
    let isOutputMuted = false;

    if (hideUserInput) {
      output = new Writable({
        write: (chunk, encoding, callback) => {
          if (!isOutputMuted) {
            process.stdout.write(chunk, encoding);
          }
          callback();
        },
      });
    }

    const rl = readline.createInterface({
      input: process.stdin,
      output,
      terminal: true,
      historySize: hideUserInput ? 0 : 30,
    });

    rl.question(`${query}:\n\n  `, (answer) => {
      rl.close();
      resolve(answer);
    });

    isOutputMuted = hideUserInput;
  });
}

export function readYaml(filename: string) {
  return parseYaml(fs.readFileSync(filename, 'utf-8'), { filename });
}

export function writeYaml(data: any, filename: string, noRefs = false) {
  const content = stringifyYaml(data, { noRefs });

  if (process.env.NODE_ENV === 'test') {
    process.stderr.write(content);
    return;
  }
  fs.writeFileSync(filename, content);
}

export function pluralize(label: string, num: number) {
  if (label.endsWith('is')) {
    [label] = label.split(' ');
    return num === 1 ? `${label} is` : `${label}s are`;
  }
  return num === 1 ? `${label}` : `${label}s`;
}

export function handleError(e: Error, ref: string) {
  if (e instanceof ResolveError) {
    process.stderr.write(
      `Failed to resolve entrypoint definition at ${ref}:\n\n  - ${e.message}.\n\n`,
    );
  } else if (e instanceof YamlParseError) {
    process.stderr.write(
      `Failed to parse entrypoint definition at ${ref}:\n\n  - ${e.message}.\n\n`,
    );
    // TODO: codeframe
  } else { // @ts-ignore
    if (e instanceof CircularJSONNotSupportedError) {
      process.stderr.write(
        red(`Detected circular reference which can't be converted to JSON.\n`) +
        `Try to use ${blue('yaml')} output or remove ${blue('--dereferenced')}.\n\n`,
      );
    } else {
      process.stderr.write(`Something went wrong when processing ${ref}:\n\n  - ${e.message}.\n\n`);
      throw e;
    }
  }
}

export function printLintTotals(totals: Totals, definitionsCount: number) {
  const ignored = totals.ignored
    ? yellow(`${totals.ignored} ${pluralize('problem is', totals.ignored)} explicitly ignored.\n\n`)
    : '';

  if (totals.errors > 0) {
    process.stderr.write(
      red(
        `❌ Validation failed with ${totals.errors} ${pluralize('error', totals.errors)}${
          totals.warnings > 0
            ? ` and ${totals.warnings} ${pluralize('warning', totals.warnings)}`
            : ''
        }.\n${ignored}`,
      ),
    );
  } else if (totals.warnings > 0) {
    process.stderr.write(
      green(`Woohoo! Your OpenAPI ${pluralize('definition is', definitionsCount)} valid. 🎉\n`),
    );
    process.stderr.write(
      yellow(`You have ${totals.warnings} ${pluralize('warning', totals.warnings)}.\n${ignored}`),
    );
  } else {
    process.stderr.write(
      green(
        `Woohoo! Your OpenAPI ${pluralize(
          'definition is',
          definitionsCount,
        )} valid. 🎉\n${ignored}`,
      ),
    );
  }

  if (totals.errors > 0) {
    process.stderr.write(
      gray(`run \`openapi lint --generate-ignore-file\` to add all problems to the ignore file.\n`),
    );
  }

  process.stderr.write('\n');
}

export function getOutputFileName(
  entrypoint: string,
  entries: number,
  output?: string,
  ext?: BundleOutputFormat,
) {
  if (!output) {
    return { outputFile: 'stdout', ext: ext || 'yaml' };
  }

  let outputFile = output;
  if (entries > 1) {
    ext = ext || (extname(entrypoint).substring(1) as BundleOutputFormat);
    if (!outputExtensions.includes(ext as any)) {
      throw new Error(`Invalid file extension: ${ext}.`);
    }
    outputFile = join(output, basename(entrypoint, extname(entrypoint))) + '.' + ext;
  } else {
    if (output) {
      ext = ext || (extname(output).substring(1) as BundleOutputFormat);
    }
    ext = ext || (extname(entrypoint).substring(1) as BundleOutputFormat);
    if (!outputExtensions.includes(ext as any)) {
      throw new Error(`Invalid file extension: ${ext}.`);
    }
    outputFile = join(dirname(outputFile), basename(outputFile, extname(outputFile))) + '.' + ext;
  }
  return { outputFile, ext };
}

export function printUnusedWarnings(config: LintConfig) {
  const { preprocessors, rules, decorators } = config.getUnusedRules();
  if (rules.length) {
    process.stderr.write(
      yellow(
        `[WARNING] Unused rules found in ${blue(config.configFile || '')}: ${rules.join(', ')}.\n`,
      ),
    );
  }
  if (preprocessors.length) {
    process.stderr.write(
      yellow(
        `[WARNING] Unused preprocessors found in ${blue(
          config.configFile || '',
        )}: ${preprocessors.join(', ')}.\n`,
      ),
    );
  }
  if (decorators.length) {
    process.stderr.write(
      yellow(
        `[WARNING] Unused decorators found in ${blue(config.configFile || '')}: ${decorators.join(
          ', ',
        )}.\n`,
      ),
    );
  }

  if (rules.length || preprocessors.length) {
    process.stderr.write(`Check the spelling and verify you added plugin prefix.\n`);
  }
}

export function exitWithError(message: string) {
  process.stderr.write(red(message)+ '\n\n');
  process.exit(1);
}

/**
 * Checks if dir is subdir of parent
 */
export function isSubdir(parent: string, dir: string): boolean {
  const relative = path.relative(parent, dir);
  return !!relative && !/^..($|\/)/.test(relative) && !path.isAbsolute(relative);
}
