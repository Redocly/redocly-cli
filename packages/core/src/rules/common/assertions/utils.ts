import { isRef } from '../../../ref-utils';
import { Problem, ProblemSeverity, UserContext } from '../../../walk';
import { asserts } from './asserts';

export type OrderDirection = 'asc' | 'desc';

export type OrderOptions = {
  direction: OrderDirection;
  property: string;
};

export type AssertToApply = {
  name: string;
  assertId?: string;
  conditions: any;
  message?: string;
  severity?: ProblemSeverity;
  suggest?: string[];
  runsOnKeys: boolean;
  runsOnValues: boolean;
};

export function buildVisitorObject(
  subject: string,
  context: Record<string, any>[],
  subjectVisitor: any,
) {
  if (!context) {
    return { [subject]: subjectVisitor };
  }

  let currentVisitorLevel: Record<string, any> = {};
  const visitor: Record<string, any> = currentVisitorLevel;

  for (let index = 0; index < context.length; index++) {
    const node = context[index];
    if (context.length === index + 1 && node.type === subject) {
      // Visitors don't work properly for the same type nested nodes, so
      // as a workaround for that we don't create separate visitor for the last element
      // which is the same as subject;
      // we will check includes/excludes it in the last visitor.
      continue;
    }
    const matchParentKeys = node.matchParentKeys;
    const excludeParentKeys = node.excludeParentKeys;

    if (matchParentKeys && excludeParentKeys) {
      throw new Error(
        `Both 'matchParentKeys' and 'excludeParentKeys' can't be under one context item`,
      );
    }

    if (matchParentKeys || excludeParentKeys) {
      currentVisitorLevel[node.type] = {
        skip: (_value: any, key: string) => {
          if (matchParentKeys) {
            return !matchParentKeys.includes(key);
          }
          if (excludeParentKeys) {
            return excludeParentKeys.includes(key);
          }
        },
      };
    } else {
      currentVisitorLevel[node.type] = {};
    }
    currentVisitorLevel = currentVisitorLevel[node.type];
  }

  currentVisitorLevel[subject] = subjectVisitor;

  return visitor;
}

export function buildSubjectVisitor(
  properties: string | string[],
  asserts: AssertToApply[],
  context?: Record<string, any>[],
) {
  return function (node: any, { report, location, key, type, resolve }: UserContext) {
    // We need to check context's last node if it has the same type as subject node;
    // if yes - that means we didn't create context's last node visitor,
    // so we need to handle 'matchParentKeys' and 'excludeParentKeys' conditions here;
    if (context) {
      const lastContextNode = context[context.length - 1];
      if (lastContextNode.type === type.name) {
        const matchParentKeys = lastContextNode.matchParentKeys;
        const excludeParentKeys = lastContextNode.excludeParentKeys;

        if (matchParentKeys && !matchParentKeys.includes(key)) {
          return;
        }
        if (excludeParentKeys && excludeParentKeys.includes(key)) {
          return;
        }
      }
    }

    if (properties) {
      properties = Array.isArray(properties) ? properties : [properties];
    }

    for (const assert of asserts) {
      if (properties) {
        for (const property of properties) {
          // we can have resolvable scalar so need to resolve value here.
          const value = isRef(node[property]) ? resolve(node[property])?.node : node[property];
          runAssertion(value, assert, location.child(property), report);
        }
      } else {
        runAssertion(Object.keys(node), assert, location.key(), report);
      }
    }
  };
}

export function getIntersectionLength(keys: string[], properties: string[]): number {
  const props = new Set(properties);
  let count = 0;
  for (const key of keys) {
    if (props.has(key)) {
      count++;
    }
  }
  return count;
}

export function isOrdered(value: any[], options: OrderOptions | OrderDirection): boolean {
  const direction = (options as OrderOptions).direction || (options as OrderDirection);
  const property = (options as OrderOptions).property;
  for (let i = 1; i < value.length; i++) {
    let currValue = value[i];
    let prevVal = value[i - 1];

    if (property) {
      if (!value[i][property] || !value[i - 1][property]) {
        return false; // property doesn't exist, so collection is not ordered
      }
      currValue = value[i][property];
      prevVal = value[i - 1][property];
    }

    const result = direction === 'asc' ? currValue >= prevVal : currValue <= prevVal;
    if (!result) {
      return false;
    }
  }
  return true;
}

function runAssertion(
  values: string | string[],
  assert: AssertToApply,
  location: any,
  report: (problem: Problem) => void,
) {
  const lintResult = asserts[assert.name](values, assert.conditions);
  if (!lintResult) {
    report({
      message: assert.message || `The ${assert.assertId} doesn't meet required conditions`,
      location,
      forceSeverity: assert.severity,
      suggest: assert.suggest,
      ruleId: assert.assertId,
    });
  }
}
