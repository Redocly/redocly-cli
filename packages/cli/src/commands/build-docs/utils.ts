import { createElement } from 'react';
import { createStore, Redoc } from 'redoc';
import { Config, isAbsoluteUrl } from '@redocly/openapi-core';

import { renderToString } from 'react-dom/server';
import { ServerStyleSheet } from 'styled-components';
import { compile } from 'handlebars';
import { join } from 'path';
import { existsSync, lstatSync, readFileSync } from 'fs';

import type { BuildDocsOptions } from './types';
import { red } from 'colorette';
import { exitWithError } from '../../utils';

export function getObjectOrJSON(
  openapiOptions: string | Record<string, unknown>,
  config: Config
): JSON | Record<string, unknown> | Config {
  switch (typeof openapiOptions) {
    case 'object':
      return openapiOptions;
    case 'string':
      try {
        if (existsSync(openapiOptions) && lstatSync(openapiOptions).isFile()) {
          return JSON.parse(readFileSync(openapiOptions, 'utf-8'));
        } else {
          return JSON.parse(openapiOptions);
        }
      } catch (e) {
        process.stderr.write(
          red(
            `Encountered error:\n\n${openapiOptions}\n\nis neither a file with a valid JSON object neither a stringified JSON object.`
          )
        );
        exitWithError(e);
      }
      break;
    default: {
      if (config) {
        process.stderr.write(`Found ${config.configFile} and using theme.openapi options`);

        return config.theme.openapi ? config.theme.openapi : {};
      }
      return {};
    }
  }
  return {};
}

export async function getPageHTML(
  api: any,
  pathToApi: string,
  {
    cdn,
    title,
    disableGoogleFont,
    templateFileName,
    templateOptions,
    redocOptions = {},
    redocCurrentVersion,
  }: BuildDocsOptions
) {
  process.stderr.write('Prerendering docs');

  const apiUrl = redocOptions.specUrl || (isAbsoluteUrl(pathToApi) ? pathToApi : undefined);
  const store = await createStore(api, apiUrl, redocOptions);
  const sheet = new ServerStyleSheet();

  const html = renderToString(sheet.collectStyles(createElement(Redoc, { store })));
  const state = await store.toJS();
  const css = sheet.getStyleTags();

  templateFileName = templateFileName ? templateFileName : join(__dirname, './template.hbs');
  const template = compile(readFileSync(templateFileName).toString());
  return template({
    redocHTML: `
      <div id="redoc">${html || ''}</div>
      <script>
      ${`const __redoc_state = ${sanitizeJSONString(JSON.stringify(state))};` || ''}

      var container = document.getElementById('redoc');
      Redoc.${'hydrate(__redoc_state, container)'};

      </script>`,
    redocHead:
      (cdn
        ? '<script src="https://cdn.redoc.ly/redoc/latest/bundles/redoc.standalone.js"></script>'
        : `<script src="https://cdn.redoc.ly/redoc/v${redocCurrentVersion}/bundles/redoc.standalone.js"></script>`) +
      css,
    title: title || api.info.title || 'ReDoc documentation',
    disableGoogleFont,
    templateOptions,
  });
}

export function sanitizeJSONString(str: string): string {
  return escapeClosingScriptTag(escapeUnicode(str));
}

// see http://www.thespanner.co.uk/2011/07/25/the-json-specification-is-now-wrong/
export function escapeClosingScriptTag(str: string): string {
  return str.replace(/<\/script>/g, '<\\/script>');
}

// see http://www.thespanner.co.uk/2011/07/25/the-json-specification-is-now-wrong/
export function escapeUnicode(str: string): string {
  return str.replace(/\u2028|\u2029/g, (m) => '\\u202' + (m === '\u2028' ? '8' : '9'));
}
