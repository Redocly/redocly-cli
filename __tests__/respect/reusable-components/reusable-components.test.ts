import { getParams, getCommandOutput } from '../utils';
import { join } from 'path';

test('should use inputs from CLI and env to mapp with resolved refs', () => {
  const indexEntryPoint = join(process.cwd(), 'packages/cli/lib/index.js');
  const fixturesPath = join(__dirname, 'reusable-components.arazzo.yaml');
  const args = getParams(indexEntryPoint, [
    'respect',
    fixturesPath,
    '--verbose',
    '--input',
    '{"store_id":"42","my_pet_tags":["one","two"]}',
    '--input',
    'reusable-test="123"',
  ]);
  const result = getCommandOutput(args);

  expect(result).toMatchSnapshot();
});
