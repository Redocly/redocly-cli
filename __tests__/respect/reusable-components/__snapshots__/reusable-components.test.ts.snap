// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should use inputs from CLI and env to map with resolved refs 1`] = `
"────────────────────────────────────────────────────────────────────────────────

  Running workflow reusable-components.arazzo.yaml / get-museum-hours

  ✓ GET /museum-hours - step get-museum-hours-step-success-action

    Request URL: https://redocly.com/_mock/demo/openapi/museum-api/museum-hours
    Request Headers:
      accept: application/json, application/problem+json
      pagesize: 256
      authorization: Basic Og==


    Response status code: 200
    Response time: <test> ms
    Response Body:
      [
        {
          "date": "2023-09-11",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-12",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-13",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-14",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-15",
          "timeOpen": "10:00",
          "timeClose": "16:00"
        },
        {
          "date": "2023-09-18",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-19",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-20",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-21",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-22",
          "timeOpen": "10:00",
          "timeClose": "16:00"
        }
      ]

    ✓ success criteria check - $statusCode == 200
    ✓ status code check - $statusCode in [200, 400, 404]
    ✓ content-type check
    ✓ schema check

  ✗ GET /museum-hours - step get-museum-hours-step-failure-action-goto

    Request URL: https://redocly.com/_mock/demo/openapi/museum-api/museum-hours
    Request Headers:
      accept: application/json, application/problem+json
      pagesize: 256
      authorization: Basic Og==


    Response status code: 200
    Response time: <test> ms
    Response Body:
      [
        {
          "date": "2023-09-11",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-12",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-13",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-14",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-15",
          "timeOpen": "10:00",
          "timeClose": "16:00"
        },
        {
          "date": "2023-09-18",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-19",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-20",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-21",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-22",
          "timeOpen": "10:00",
          "timeClose": "16:00"
        }
      ]

    ✗ success criteria check - $statusCode == 201
    ✓ status code check - $statusCode in [200, 400, 404]
    ✓ content-type check
    ✓ schema check

  Running failure action get-museum-hours-failure-action for the step get-museum-hours-step-failure-action-goto
  Running workflow reusable-components.arazzo.yaml / final-workflow

  ✓ GET /museum-hours - step final-workflow-step

    Request URL: https://redocly.com/_mock/demo/openapi/museum-api/museum-hours
    Request Headers:
      accept: application/json, application/problem+json
      pagesize: 256
      authorization: Basic Og==


    Response status code: 200
    Response time: <test> ms
    Response Body:
      [
        {
          "date": "2023-09-11",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-12",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-13",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-14",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-15",
          "timeOpen": "10:00",
          "timeClose": "16:00"
        },
        {
          "date": "2023-09-18",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-19",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-20",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-21",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-22",
          "timeOpen": "10:00",
          "timeClose": "16:00"
        }
      ]

    ✓ success criteria check - $statusCode == 200
    ✓ status code check - $statusCode in [200, 400, 404]
    ✓ content-type check
    ✓ schema check


────────────────────────────────────────────────────────────────────────────────

  Running workflow reusable-components.arazzo.yaml / events-crud

  ✓ GET /special-events - step list-events

    Request URL: https://redocly.com/_mock/demo/openapi/museum-api/special-events
    Request Headers:
      accept: application/json, application/problem+json
      pagesize: 256
      authorization: Basic Og==
      page: 1


    Response status code: 200
    Response time: <test> ms
    Response Body:
      [
        {
          "eventId": "f3e0e76e-e4a8-466e-ab9c-ae36c15b8e97",
          "name": "Sasquatch Ballet",
          "location": "Seattle... probably",
          "eventDescription": "They're big, they're hairy, but they're also graceful. Come learn how the biggest feet can have the lightest touch.",
          "dates": [
            "2023-12-15",
            "2023-12-22"
          ],
          "price": 40
        },
        {
          "eventId": "2f14374a-9c65-4ee5-94b7-fba66d893483",
          "name": "Solar Telescope Demonstration",
          "location": "Far from the sun.",
          "eventDescription": "Look at the sun without going blind!",
          "dates": [
            "2023-09-07",
            "2023-09-14"
          ],
          "price": 50
        },
        {
          "eventId": "6aaa61ba-b2aa-4868-b803-603dbbf7bfdb",
          "name": "Cook like a Caveman",
          "location": "Fire Pit on East side",
          "eventDescription": "Learn to cook on an open flame.",
          "dates": [
            "2023-11-10",
            "2023-11-17",
            "2023-11-24"
          ],
          "price": 5
        },
        {
          "eventId": "602b75e1-5696-4ab8-8c7a-f9e13580f910",
          "name": "Underwater Basket Weaving",
          "location": "Rec Center Pool next door.",
          "eventDescription": "Learn to weave baskets underwater.",
          "dates": [
            "2023-09-12",
            "2023-09-15"
          ],
          "price": 15
        },
        {
          "eventId": "dad4bce8-f5cb-4078-a211-995864315e39",
          "name": "Mermaid Treasure Identification and Analysis",
          "location": "Room Sea-12",
          "eventDescription": "Join us as we review and classify a rare collection of 20 thingamabobs, gadgets, gizmos, whoosits, and whatsits — kindly donated by Ariel.",
          "dates": [
            "2023-09-05",
            "2023-09-08"
          ],
          "price": 30
        },
        {
          "eventId": "6744a0da-4121-49cd-8479-f8cc20526495",
          "name": "Time Traveler Tea Party",
          "location": "Temporal Tearoom",
          "eventDescription": "Sip tea with important historical figures.",
          "dates": [
            "2023-11-18",
            "2023-11-25",
            "2023-12-02"
          ],
          "price": 60
        },
        {
          "eventId": "3be6453c-03eb-4357-ae5a-984a0e574a54",
          "name": "Pirate Coding Workshop",
          "location": "Computer Room",
          "eventDescription": "Captain Blackbeard shares his love of the C...language. And possibly Arrrrr (R lang).",
          "dates": [
            "2023-10-29",
            "2023-10-30",
            "2023-10-31"
          ],
          "price": 45
        },
        {
          "eventId": "9d90d29a-2af5-4206-97d9-9ea9ceadcb78",
          "name": "Llama Street Art Through the Ages",
          "location": "Auditorium",
          "eventDescription": "Llama street art?! Alpaca my bags -- let's go!",
          "dates": [
            "2023-10-29",
            "2023-10-30",
            "2023-10-31"
          ],
          "price": 45
        },
        {
          "eventId": "a3c7b2c4-b5fb-4ef7-9322-00a919864957",
          "name": "The Great Parrot Debate",
          "location": "Outdoor Amphitheatre",
          "eventDescription": "See leading parrot minds discuss important geopolitical issues.",
          "dates": [
            "2023-11-03",
            "2023-11-10"
          ],
          "price": 35
        },
        {
          "eventId": "b92d46b7-4c5d-422b-87a5-287767e26f29",
          "name": "Eat a Bunch of Corn",
          "location": "Cafeteria",
          "eventDescription": "We accidentally bought too much corn. Please come eat it.",
          "dates": [
            "2023-11-10",
            "2023-11-17",
            "2023-11-24"
          ],
          "price": 5
        }
      ]

    ✓ status code check - $statusCode in [200, 400, 404]
    ✓ content-type check
    ✓ schema check

  ✓ POST /special-events - step create-event

    Request URL: https://redocly.com/_mock/demo/openapi/museum-api/special-events
    Request Headers:
      content-type: application/json
      accept: application/json, application/problem+json
      pagesize: 256
      authorization: Basic Og==
      page: 42
    Request Body:
      {
        "reusable": "\\"123\\"",
        "store_id": "42",
        "my_pet_tags": [
          "one",
          "two"
        ],
        "name": "Mermaid Treasure Identification and Analysis",
        "location": "Under the seaaa 🦀 🎶 🌊.",
        "eventDescription": "Join us as we review and classify a rare collection of 20 thingamabobs, gadgets, gizmos, whoosits, and whatsits, kindly donated by Ariel.",
        "dates": [
          "2023-09-05",
          "2023-09-08"
        ],
        "price": 0
      }


    Response status code: 201
    Response time: <test> ms
    Response Body:
      {
        "eventId": "dad4bce8-f5cb-4078-a211-995864315e39",
        "name": "Mermaid Treasure Identification and Analysis",
        "location": "Under the seaaa 🦀 🎶 🌊.",
        "eventDescription": "Join us as we review and classify a rare collection of 20 thingamabobs, gadgets, gizmos, whoosits, and whatsits, kindly donated by Ariel.",
        "dates": [
          "2023-09-05",
          "2023-09-08"
        ],
        "price": 0
      }

    ✓ success criteria check - $statusCode == 201
    ✓ status code check - $statusCode in [201, 400, 404]
    ✓ content-type check
    ✓ schema check

────────────────────────────────────────────────────────────────────────────────

  Running workflow reusable-components.arazzo.yaml / final-workflow

  ✓ GET /museum-hours - step final-workflow-step

    Request URL: https://redocly.com/_mock/demo/openapi/museum-api/museum-hours
    Request Headers:
      accept: application/json, application/problem+json
      pagesize: 256
      authorization: Basic Og==


    Response status code: 200
    Response time: <test> ms
    Response Body:
      [
        {
          "date": "2023-09-11",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-12",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-13",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-14",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-15",
          "timeOpen": "10:00",
          "timeClose": "16:00"
        },
        {
          "date": "2023-09-18",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-19",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-20",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-21",
          "timeOpen": "09:00",
          "timeClose": "18:00"
        },
        {
          "date": "2023-09-22",
          "timeOpen": "10:00",
          "timeClose": "16:00"
        }
      ]

    ✓ success criteria check - $statusCode == 200
    ✓ status code check - $statusCode in [200, 400, 404]
    ✓ content-type check
    ✓ schema check


  Failed tests info:

  Workflow name: get-museum-hours

    stepId - get-museum-hours-step-failure-action-goto
    ✗ success criteria check
      Checking simple criteria: {"condition":"$statusCode == 201"}
      
  Summary for reusable-components.arazzo.yaml
  
  Workflows: 2 passed, 1 failed, 3 total
  Steps: 5 passed, 1 failed, 6 total
  Checks: 22 passed, 1 failed, 23 total
  Time: <test>ms


┌─────────────────────────────────────────────────────────────────────────┬────────────┬─────────┬─────────┬──────────┐
│ Filename                                                                │ Workflows  │ Passed  │ Failed  │ Warnings │
├─────────────────────────────────────────────────────────────────────────┼────────────┼─────────┼─────────┼──────────┤
│ x reusable-components.arazzo.yaml                                       │ 3          │ 2       │ 1       │ -        │
└─────────────────────────────────────────────────────────────────────────┴────────────┴─────────┴─────────┴──────────┘


 Tests exited with error 

"
`;
